src/symmetric.rs

```diff
@@ -0,0 +1,405 @@
+    pub fn new(cipher: Cipher, key: &[u8], rounds: Option<u32>) -> Result<Self> {
+        unsafe {
+            let mut raw = mem::uninitialized();
+            tryt!(ffi::ecb_start(cipher.index(), key.as_ptr(), key.len() as c_int, rounds.unwrap_or(0) as c_int, &mut raw));
+
+            Ok(ECB(raw))
+        }
+    }
+}
+
+impl CipherMode for ECB {
+    unsafe fn encrypt_unchecked(&mut self, plaintext: &[u8], ciphertext: &mut [u8]) -> Result<()> {
+        tryt!(ffi::ecb_encrypt(plaintext.as_ptr(), ciphertext.as_mut_ptr(), plaintext.len() as u64, &mut self.0));
+
+        Ok(())
+    }
+
+    unsafe fn decrypt_unchecked(&mut self, ciphertext: &[u8], plaintext: &mut [u8]) -> Result<()> {
+        tryt!(ffi::ecb_decrypt(ciphertext.as_ptr(), plaintext.as_mut_ptr(), ciphertext.len() as u64, &mut self.0));
+
+        Ok(())
+    }
+}
+
+impl Drop for ECB {
+    fn drop(&mut self) {
+        unsafe {
+            ffi::ecb_done(&mut self.0);
+        }
+    }
+}
+
+
+/// CBC or Cipher Block Chaining mode is a simple mode designed to prevent trivial forms of replay and swap attacks on
+/// ciphers.
+///
+/// It is important that the initialization vector be unique and preferably random for each message encrypted under the
+/// same key.
+pub struct CBC(ffi::symmetric_CBC);
+
+impl CBC {
+    pub fn new(cipher: Cipher, iv: &[u8], key: &[u8], rounds: Option<u32>) -> Result<Self> {
+        // Validate the IV size since LibTomCrypt doesn't.
+        if iv.len() != cipher.block_size() {
+            return Err(Error::from_code(ffi::CRYPT_INVALID_ARG));
+        }
+
+        unsafe {
+            let mut raw = mem::uninitialized();
+            tryt!(ffi::cbc_start(cipher.index(), iv.as_ptr(), key.as_ptr(), key.len() as c_int, rounds.unwrap_or(0) as c_int, &mut raw));
+
+            Ok(CBC(raw))
+        }
+    }
+}
+
+impl CipherMode for CBC {
+    unsafe fn encrypt_unchecked(&mut self, plaintext: &[u8], ciphertext: &mut [u8]) -> Result<()> {
+        tryt!(ffi::cbc_encrypt(plaintext.as_ptr(), ciphertext.as_mut_ptr(), plaintext.len() as u64, &mut self.0));
+
+        Ok(())
+    }
+
+    unsafe fn decrypt_unchecked(&mut self, ciphertext: &[u8], plaintext: &mut [u8]) -> Result<()> {
+        tryt!(ffi::cbc_decrypt(ciphertext.as_ptr(), plaintext.as_mut_ptr(), ciphertext.len() as u64, &mut self.0));
+
+        Ok(())
+    }
+}
+
+impl Drop for CBC {
+    fn drop(&mut self) {
+        unsafe {
+            ffi::cbc_done(&mut self.0);
+        }
+    }
+}
+
+
+/// CTR or Counter Mode is a mode which only uses the encryption function of the cipher.
+///
+/// As long as the initialization vector is random for each message encrypted under the same key replay and swap attacks
+/// are infeasible. CTR mode may look simple but it is as secure as the block cipher is under a chosen plaintext attack
+/// (provided the initialization vector is unique).
+pub struct CTR(ffi::symmetric_CTR);
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum CTRCounterMode {
```

Comment:
This should probably be `CTRMode`, `CounterMode` or `CTREndianess` CTRCounter is a bit redundant ;)

Reply:
Hmm, yeah good point. I like `CTREndianess`, it explains what it means better.